### PR TITLE
Add flexibility to ESRP signing and scanning tasks

### PR DESCRIPTION
**Pull request by spaceisfun**
- Improved logging.
- Param to use custom signing JSON
- Param to run signtool /verify after signing
- Param to enable/disable page hash signing
- Param to use different ESRP scan service name

### CODE REVIEW

File: template-compliance/custom-sign.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+parameters:
+  - name: "signOutputPath"
+    default: "$(Build.ArtifactStagingDirectory)\\signed"
+
+steps:
+- pwsh: |
+    if ($null -eq $env:ESRP_TEMPLATE_CUSTOM_JSON)
+    {
+      Write-Error "ESRP_TEMPLATE_CUSTOM_JSON variable must be set!"
```
Comment (TravisEz13) [inline template-compliance/custom-sign.yml:9]: Add example to readme

File: template-compliance/authenticode-sign.yml [hunk 2]
```diff
@@ -7,6 +7,10 @@ parameters:
     default: "*.dll,*.exe"
   - name: "certificateId"
     default: "CP-230012"
+  - name: "verifySignature"
```
Comment (TravisEz13) [inline template-compliance/authenticode-sign.yml:10]: add examples in readme

File: EsrpScan.yml [hunk 3]
```diff
@@ -3,17 +3,25 @@ parameters:
     default: "$(Pipeline.Workspace)"
   - name: "pattern"
     default: "*.dll,*.exe"
+  - name: "scanningService"
```
Comment (TravisEz13) [inline EsrpScan.yml:6]: add example in readme

**Review by TravisEz13 — CHANGES_REQUESTED** [review]

File: EsrpScan.yml (hunk 3 above)
Comment (spaceisfun) [reply EsrpScan.yml:6]: Added

File: template-compliance/authenticode-sign.yml (hunk 2 above)
Comment (spaceisfun) [reply template-compliance/authenticode-sign.yml:10]: Added

**Review by TravisEz13 — APPROVED** [review]

**Comment by TravisEz13** [comment]
@JamesWTruher Please review

File: README.md [hunk 4]
```diff
@@ -158,6 +182,77 @@ This example signs `pkg` files recursively, using minimatch.
...
+This example uses a custom signing (Azure DevOps) service connection name.
+
+```yaml
+  - template: EsrpSign.yml@ComplianceRepo
+      parameters:
+        buildOutputPath: $(signSrcPath)
+        signOutputPath: $(signOutPath)
+        certificateId: "CP-230012"
+        pattern: |
+          **\*.dll
```
Comment (JamesWTruher) [inline README.md:249]: does this pattern need minimatch?

**Review by JamesWTruher — COMMENTED** [review]
just a couple of questions

File: EsrpSign.yml [hunk 5]
```diff
@@ -27,6 +33,18 @@ steps:
...
+    Write-Verbose -Verbose "signingService = '${{ parameters.signingService }}'"
+    Write-Verbose -Verbose "shouldSign = '${{ parameters.shouldSign }}'"
+    Write-Verbose -Verbose "alwaysCopy = '${{ parameters.alwaysCopy }}'"
+    Write-Verbose -Verbose "useCustomEsrpJson = '${{ parameters.useCustomEsrpJson }}'"
+    Write-Verbose -Verbose "verifySignature = '${{ parameters.verifySignature }}'"
+    Write-Verbose -Verbose "pageHash = '${{ parameters.pageHash }}'"
+
+    if (('${{ parameters.useCustomEsrpJson }}' -ne 'false') -and ('${{ parameters.certificateId }}' -ne ''))
+    {
+      Write-Error "Only one of useCustomEsrpJson and certificateId must be set!"
```
[the remark this answers is not in the document]
Comment (TravisEz13) [reply EsrpSign.yml:46]: It should be a throw.
```suggestion
      throw "Only one of useCustomEsrpJson and certificateId must be set!"
```

File: README.md (hunk 4 above)
Comment (TravisEz13) [reply README.md:249]: it does

Comment (spaceisfun) [reply README.md:249]: good catch, fixed

**Pull request merged**